src/som/interpreter/SArguments.java

```diff
@@ -58,4 +91,139 @@ public static SImmutableArray getArgumentsWithoutReceiver(final Object[] argumen
+      entryLoad.loadShadowStackEntry(result, expression, frame, false);
+    }
+    return result;
+  }
+
+  public static Object[] getPlainNoArgumentsWithReceiver(final Object receiver,
+      final ExpressionNode expression,
+      final ShadowStackEntryLoad entryLoad,
+      final VirtualFrame frame) {
+    if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      Object[] arguments = new Object[] {receiver, null};
+      entryLoad.loadShadowStackEntry(arguments, expression, frame, false);
+      return arguments;
+    } else {
+      return new Object[] {receiver};
+    }
+  }
+
+  public static Object[] getPlain1ArgumentWithReceiver(final Object receiver,
+      final Object firstArg,
+      final ExpressionNode expression,
+      final ShadowStackEntryLoad entryLoad,
+      final VirtualFrame frame) {
+    if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      Object[] arguments = new Object[] {receiver, firstArg, null};
+      entryLoad.loadShadowStackEntry(arguments, expression, frame, false);
+      return arguments;
+    } else {
+      return new Object[] {receiver, firstArg};
+    }
+  }
+
+  public static Object[] getPlain2ArgumentsWithReceiver(final Object receiver,
+      final Object firstArg,
+      final Object secondArg,
+      final ExpressionNode expression,
+      final ShadowStackEntryLoad entryLoad,
+      final VirtualFrame frame) {
+    if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      Object[] arguments = new Object[] {receiver, firstArg, secondArg, null};
+      entryLoad.loadShadowStackEntry(arguments, expression, frame, false);
+      return arguments;
+    } else {
+      return new Object[] {receiver, firstArg, secondArg};
+    }
+  }
+
+  public static Object[] getPlainArguments(final Object[] args) {
+    if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      Object[] newArgs = new Object[args.length + 1];
+      for (int i = 0; i < args.length; i++) {
+        newArgs[i] = args[i];
+      }
+      return newArgs;
+    } else {
+      return args;
+    }
+  }
+
+  public static Object[] getPromiseCallbackArgumentArray(final SBlock callback) {
+    if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      return new Object[] {callback, null, null};
+    } else {
+      return new Object[] {callback, null};
+    }
+  }
+
+  public static void setShadowStackEntryWithCache(final Object[] arguments,
+      final Node expression,
+      final ShadowStackEntryLoad entryLoad,
+      final VirtualFrame frame,
+      final boolean async) {
+    if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      entryLoad.loadShadowStackEntry(arguments, expression, frame, async);
+    }
+  }
+
+  public static ShadowStackEntry instantiateTopShadowStackEntry(final Node expr) {
+    return ShadowStackEntry.createTop(expr);
+  }
+
+  public static ShadowStackEntry instantiateShadowStackEntry(final ShadowStackEntry previous,
+      final Node expr, final boolean async) {
+    if (async) {
```

Comment:
Could you add the following? This is a compile time check that async is known and can be used for optimization. It's not a hint that this is the case, it will only result in an error if it isn't a constant value. Good for debugging of performance issues.

```suggestion
    CompilerAsserts.partialEvaluationConstant(async);
    if (async) {
```

Reply:
Done, it works.